### ISSUE TITLE
Swap chain surface size on retina displays

### ISSUE THREAD

**Issue by DiligentGraphics** [KhronosGroup/MoltenVK#428]
On retina display, `vkGetPhysicalDeviceSurfaceCapabilitiesKHR()` reports surfaces size in logical points rather than pixels, resulting in image being inappropriately downscaled. I was unable to find any way to get the actual surface size within Vulkan, so the only way that gives correct surface dimensions is to ignore `surfCapabilities.currentExtent` and rely on
```
NSRect viewRectPixels = [view convertRectToBacking:viewRectPoints];
```

**Comment by KenThomases** [comment]
Are you supplying MoltenVK with a view (from which it obtains a layer) or with a layer directly? Either way, are you using `MTKView` to provide and configure the layer or are you setting up the layer yourself? If the latter, how?  In particular, are you setting its `contentsScale`? Or, put another way, what is the layer's `drawableSize`?

**Comment by DiligentGraphics** [comment]
Thank you for super fast response! 
I am using `MTKView` as follows:

```Objective-c
- (nonnull instancetype)initWithMetalKitView:(nonnull MTKView *)mtkView
{
    self = [super init];
    if(self)
    {
        _theApp.reset(CreateApplication());
        _theApp->Initialize(mtkView);
    }

    return self;
}

- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size
{
    NSRect viewRectPoints = [view bounds];
    NSRect viewRectPixels = [view convertRectToBacking:viewRectPoints];
    
    _theApp->WindowResize(viewRectPixels.size.width, viewRectPixels.size.height);
}
```
So I supply `MTKView` to the swap chain creation code. When `drawableSizeWillChange` is called, `size` contains logical size (800x600) and this is also what `vkGetPhysicalDeviceSurfaceCapabilitiesKHR()` reports. `viewRectPixels` is however 1600x1200 and this is the actual surface size I need.

I am not a MacOS/iOS developer, and I took the code above from Apple's Metal example, so I may be doing something wrong. I do not set `contentsScale` anywhere in the code. Shouldn't it be handled by the system?

**Comment by DiligentGraphics** [comment]
When `drawableSizeWillChange` is called, the `size` parameter is `(1600, 1200)`. Swap chain extents, however, are reported as `(800, 600)`.
If I manually set `contentsScale` for the layer:

```Objective-c
_view = (MTKView *)self.view;
auto layer = [_view layer];
layer.contentsScale = 2.0;
```

the extents are reported correctly.

I never touch neither `drawableSize` nor `autoResizeDrawable`.

**Comment by tokyovigilante** [comment]
I think this is as designed, the default `contentsScale` for a CAMetalLayer is 1.0. Setting it to 2.0 enables HiDPI/Retina support as you've found.

**Comment by KenThomases** [comment]
Hmm.  The docs for `drawableSize` claim that it returns the `bounds` * `contentsScale`. But @DiligentGraphics reports that the delegate method is getting a size that's scaled by 2.0 (as expected). So, there's some sort of problem somewhere, perhaps only in the docs.

But I'm suspicious. For example, why is the delegate method being called at all? I suspect it happens when the view actually ends up on screen and only then is it determined that it's on a Retina display. Before that, the `contentsScale` was 1.0 and the `drawableSize` was 1x the view's size in points (not pixels).

There's a thing where MoltenVK requires that the swapchain be recreated to get the new `drawableSize` from the layer. So, I think that one needs to do that and @DiligentGraphics  isn't. (I'm a bit out of my depth on the Vulkan parts of this.)

**Comment by DiligentGraphics** [comment]
@KenThomases  `drawableSize`  indeed returns the `bounds * contentsScale`. My problem was that on retina display, `contentsScale` of the metal layer had to be explicitly set to 2.0 by the application. That was not obvious at all, especially for someone who does not have a lot of Mac experience. I copied the code from Apple's Metal example that did not do that.

The swap chain is recreated in `drawableSizeWillChange` delegate:

```Objective-C
/// Called whenever view changes orientation or is resized
- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size
{
    if (_theApp)
    {
        _theApp->WindowResize(size.width, size.height);
    }
}
```

Before setting the content scale, after the resize the size of swap chain was consistently quarter of the screen.

This is the solution that seems to be working fine:

```Objective-C
- (void)viewDidLoad
{
    [super viewDidLoad];

    _view = (MTKView *)self.view;
    NSScreen* screen = [NSScreen mainScreen];
    auto scaleFactor = [screen backingScaleFactor];
    auto layer = [_view layer];
    layer.contentsScale = scaleFactor;

    _renderer = [[MTKRenderer alloc] initWithMetalKitView:_view];
    _theApp = [_renderer getApp];
   ...
}
```

@tokyovigilante the `drawableSize` is returned in native pixels, as claimed. The problem was that before I set `contentsScale`, `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` was reporting the size in logical points.

**Comment by tokyovigilante** [comment]
> The problem was that before I set contentsScale, vkGetPhysicalDeviceSurfaceCapabilitiesKHR was reporting the size in logical points, and was not consistent with drawableSize

Understood, however I think what it actually implies is that without setting `contentsScale` your renderer surface is drawn at 1x resolution then scaled up on a Retina display, rather than drawing at native resolution. So for a 400x200 window on a Retina display, you get a 400x200 drawable that is then scaled by 2x (by macOS) to fit the native 800x400 pixel window. But setting it to 2.0 gives you a 800x400 drawable that doesn't require further scaling.

You can use this to your advantage, say setting it to 0.5 for a further speedup at the cost of image quality. Conversely setting it to 2.0 (or whatever NSScreen.backingScaleFactor()` reports will render at native resolution but will be slower.

**Comment by tokyovigilante** [comment]
Hmm, that definitely seems like an MVK bug then. Using Metal directly, the `contentsScale` just controls the relative resolution of the layer, the actual rendered image is always scaled as required to fit the view (ie 100% layer coverage), Have you tried using a CAMetalLayer directly and just adding it to a layer-backed view?

**Comment by DiligentGraphics** [comment]
I tried to start with LunarG sample, but it [passes control over to native implementation](https://github.com/KhronosGroup/MoltenVK/blob/master/Demos/LunarG-VulkanSamples/API-Samples/macOS/DemoViewController.mm#L58) that runs the main loop, which I don't like. I was not able to make `updateLayer` called by the system, so I switched to Apple's example that uses `MTKView`. I am not used to mac/ios development, so every single thing requires a lot of googling, and I was happy that Apple's way worked.

I am surprised that nobody has reported that problem before, so I may totally be doing something not quite right. Though everything seems to be working great now.
